src/Confluent.Kafka/Impl/SafeKafkaHandle.cs

```diff
@@ -398,42 +404,65 @@ internal void Assign(ICollection<TopicPartitionOffset> partitions)
             }
             if (err != ErrorCode.NO_ERROR)
             {
-                throw new KafkaException(err, "Failed to assign partitions");
+                throw new KafkaException(err);
             }
         }
 
-        internal void Commit()
+        /// <summary>
+        ///  Dummy commit callback that does nothing but prohibits
+        ///  triggering the global offset_commit_cb.
+        ///  Used by manual commits.
+        /// </summary>
+        static void dummyOffsetCommitCb (IntPtr rk, ErrorCode err, IntPtr offsets, IntPtr opaque)
         {
-            ErrorCode err = LibRdKafka.commit(handle, IntPtr.Zero, false);
-            if (err != ErrorCode.NO_ERROR)
-            {
-                throw new KafkaException(err, "Failed to commit offsets");
-            }
+            return;
         }
 
-        internal void Commit(ICollection<TopicPartitionOffset> offsets)
+        /// <summary>
+        /// Manual sync commit, will block indefinately.
+        /// </summary>
+        /// <param name="offsets">Offsets to commit, or null for current assignment.</param>
+        /// <returns>CommittedOffsets with global or per-partition errors.</returns>
+        private CommittedOffsets commitSync (ICollection<TopicPartitionOffset> offsets)
         {
-            IntPtr list = LibRdKafka.topic_partition_list_new((IntPtr) offsets.Count);
-            if (list == IntPtr.Zero)
-            {
-                throw new Exception("Failed to create offset commit list");
-            }
-            foreach (var offset in offsets)
+            // Create temporary queue so we can get the offset commit results
+            // as an event instead of a callback.
+            // We still need to specify a dummy callback (that does nothing)
+            // to prevent the global offset_commit_cb to kick in.
+            IntPtr cQueue = LibRdKafka.queue_new(handle);
+            IntPtr cOffsets = GetCTopicPartitionList(offsets);
+            ErrorCode err = LibRdKafka.commit_queue(handle, cOffsets, cQueue, dummyOffsetCommitCb, IntPtr.Zero);
+            if (cOffsets != IntPtr.Zero)
+                LibRdKafka.topic_partition_list_destroy(cOffsets);
+            if (err != ErrorCode.NO_ERROR)
             {
-                IntPtr ptr = LibRdKafka.topic_partition_list_add(list, offset.Topic, offset.Partition);
-                Marshal.WriteInt64(
-                    ptr,
-                    (int) Marshal.OffsetOf<rd_kafka_topic_partition>("offset"),
-                    offset.Offset);
+                LibRdKafka.queue_destroy(cQueue);
+                return new CommittedOffsets(new Error(err));
             }
-            ErrorCode err = LibRdKafka.commit(handle, list, false);
-            LibRdKafka.topic_partition_list_destroy(list);
-            if (err != ErrorCode.NO_ERROR)
+
+            // Wait for commit to finish
+            IntPtr rkev = LibRdKafka.queue_poll(cQueue, -1/*infinite*/);
```

Comment:
I think having an infinite timeout here won't cut it - when the consumer is disposed this needs to be able to be cancelled (search for `CancellationToken` for examples of how to do this).

Also, do we need a timeout variant of `ConsumerAsync`?